perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ResourcesManagerBlImpl.java

```diff
@@ -344,50 +346,55 @@ public boolean isGroupAssigned(PerunSession sess, Group group, Resource resource
 	public void assignGroupToResource(PerunSession sess, Group group, Resource resource) throws InternalErrorException, WrongAttributeValueException, WrongReferenceAttributeValueException, GroupAlreadyAssignedException {
-		Vo groupVo = getPerunBl().getGroupsManagerBl().getVo(sess, group);
+		assignGroupsToResource(sess, Collections.singletonList(group), resource);
+	}
 
-		// Check if the group and resource belongs to the same VO
-		if (!groupVo.equals(this.getVo(sess, resource))) {
-			throw new InternalErrorException("Group " + group + " and resource " + resource + " belongs to the different VOs");
-		}
+	@Override
+	public void assignGroupsToResource(PerunSession perunSession, List<Group> groups, Resource resource) throws InternalErrorException, WrongAttributeValueException, WrongReferenceAttributeValueException, GroupAlreadyAssignedException {
+		Set<Member> members = new HashSet<>();
 
-		if(isGroupAssigned(sess, group, resource)) throw new GroupAlreadyAssignedException(group);
+		for(Group g: groups) {
+			Vo groupVo = getPerunBl().getGroupsManagerBl().getVo(perunSession, g);
 
-		//first we must assign group to resource and then set na check attributes, because methods checkAttributesSemantics and fillAttribute need actual state to work correctly
-		getResourcesManagerImpl().assignGroupToResource(sess, group, resource);
-		getPerunBl().getAuditer().log(sess, new GroupAssignedToResource(group, resource));
+			// Check if the group and resource belongs to the same VO
+			if (!groupVo.equals(this.getVo(perunSession, resource))) {
+				throw new InternalErrorException("Group " + g + " and resource " + resource + " belongs to the different VOs");
+			}
 
-		// if there are no assigned services, no attributes have to be checked or filled
-		if (getAssignedServices(sess, resource).isEmpty()) {
-			return;
-		}
-		// get/fill/set all required group and group-resource attributes
-		try {
-			List<Attribute> attributes = getPerunBl().getAttributesManagerBl().getResourceRequiredAttributes(sess, resource, resource, group, true);
-			attributes = getPerunBl().getAttributesManagerBl().fillAttributes(sess, resource, group, attributes, true);
-			getPerunBl().getAttributesManagerBl().setAttributes(sess, resource, group, attributes, true);
-		} catch(WrongAttributeAssignmentException | GroupResourceMismatchException ex) {
-			throw new ConsistencyErrorException(ex);
+			if(isGroupAssigned(perunSession, g, resource)) throw new GroupAlreadyAssignedException(g);
+
+			//first we must assign group to resource and then set na check attributes, because methods checkAttributesSemantics and fillAttribute need actual state to work correctly
+			getResourcesManagerImpl().assignGroupToResource(perunSession, g, resource);
+			getPerunBl().getAuditer().log(perunSession, new GroupAssignedToResource(g, resource));
+
+			// if there are no assigned services, no attributes have to be checked or filled
+			if (getAssignedServices(perunSession, resource).isEmpty()) {
```

Comment:
I would probably resolve this once, before iterating over the groups. So either attributes are processed for all groups or none of them.

Reply:
Ok, I moved the check outside the loop.